### ISSUE TITLE
Receive status update in background

### ISSUE THREAD

**Issue by tomasweigenast** [781flyingdutchman/background_downloader#341]
If the package is able to record events that happened when the app was in background or terminated and resume them using 
```dart
await _uploader.trackTasks();
await _uploader.resumeFromBackground();
```
Why we can't setup callbacks to directly receive those events on Dart, executing another Isolate obviously. Something like Firebase Messaging does for background notifications, which allows you to setup a callback that gets called although the app is in background/terminated state.

**Comment by 781flyingdutchman** [comment]
Fair question. It would be possible to set it up that way, but if you wake up the app for every background event of the downloader (typically including progress updates that happen very frequently) then the app isn't really running in the background anymore, which is the whole point of background downloading.

Can you elaborate on the use case you'd have for this (other than avoiding a few extra lines on wakeup)?

Let me look into if it's possible to configure a wakeup for final state - but even then you'd still have to process what happened when the app was truly in the background.

**Comment by tomasweigenast** [comment]
I saw the source code and the package does nothing when the app is terminated (because it saves the state to SharedPreferences, on Android, I did not see code to awake the app when WorkManager fires events), correct me if I'm wrong.

I don't mean to awake the app for every possible event (could be left to the developer). In my use case, the app must receive an event when a background **upload** is finished, even if the app was terminated because it needs to call an HTTP endpoint.

Currently, I'll try to solve that requirement using the `background_fetch` package (https://pub.dev/packages/background_fetch) which runs a callback every 15 min. approx and then check if the task was completed or not.

**Comment by 781flyingdutchman** [comment]
Correct, right now the plugin does not wake up the app. What I was suggesting is that perhaps I could add an option to do that only for a final state. That would presumably solve your issue, but I have to think about this a bit.

**Comment by tomasweigenast** [comment]
Yes, I think implementing that for a final state is good. I don't know if you can set up "hooks" for WorkManager or JobManager to call a function when the upload/download finishes even if the app is terminated.

**Comment by tomasweigenast** [comment]
@781flyingdutchman any updates on this?

**Comment by 781flyingdutchman** [comment]
No updates - it's a major feature and I haven't prioritized it at this moment.

**Comment by 781flyingdutchman** [comment]
Have a look at #321 which describes an implementation (on a separate branch) of hooks that can be configured to be called before the task starts, or after it finishes, and wakes up the app if needed.  I really need someone to test that functionality thoroughly, and it would certainly accomplish what you are looking for, so if you can take a look and let me know what you think that would be much appreciated.

**Comment by tomasweigenast** [comment]
Thank you for considering my request. I will give it a try when I can and let you know.

**Comment by 781flyingdutchman** [comment]
Published in V8.9.0 using onTaskFinished "native" callback
